Final/SyncLinkedList.h

```diff
@@ -0,0 +1,31 @@
+#ifndef SYNC_LINKED_LIST_H
+#define SYNC_LINKED_LIST_H
+
+#include <shared_mutex>
+#include <mutex>
+#include <functional>
+#include "LinkedList.h"
+
+class LinkedList;
+
+class SyncLinkedList {
```

Comment:
similar to the above comment in the linked list class, let's add this to the `dlpf` namespace so we don't pollute the global namespace, but in addition, let's also make this a new `sync` sub namespace, like so. this way, we can just rename this class `linked_list`, because in reality, it looks and functions just like a linked list, and instead let the namespace imply that it is synchronous. well named namespaces not only provide isolation/encapsulation, but can also be great for imparting semantic meaning:

```
#ifndef SYNC_LINKED_LIST_H
#define SYNC_LINKED_LIST_H

#include <shared_mutex>
#include <mutex>
#include <functional>
#include "LinkedList.h"

namespace dlpf {
	class linked_list;
	typedef struct node node;
} // dlpf

namespace dlpf::sync {
	class linked_list {
		public:
	    // Constructor
	    linked_list();
	    // Destructor
	    ~linked_list();

	    // Pushes node to front of list
	    void push(int data);
	    // Removes the first instance of a node with the given data
	    void remove(int data);

	    // Exposed to allow for easy iteration through the linked list
	    // If the iter handler returns false, the iteration will stop.
	    void for_each(std::function<bool (dlpf::node *)> iter);
		private:
  		dlpf::linked_list linked_list;
  		std::shared_mutex rwm; 
	}
} // dlpf::sync

#endif // SYNC_LINKED_LIST_H
```
note how since we put our linked_list inside the `dlpf` namespace, we also have to forward declare it within the namespace, e.g.:

```
namespace dlpf {
	class linked_list;
	typedef struct node node;
} // dlpf
```
after doing the above, we will now have a `dlpf::linked_list` and a `dlpf::sync::linked_list`. note how they both have the same name, but their functionality is distinguished via the namespace.